theano/compile/builders.py

```diff
@@ -115,40 +73,73 @@ def __hash__(self):
         # TODO: use internal variables in hash
         return hash(type(self))
 
+    def grad(self, inputs, output_grads):
+        if self.grad_op_is_cached:
+            return self.grad_op(inputs, output_grads)
+
+        if self.grad_op is None:
+            self.grad_op = []
+
+        # we need to convert a list into a single funtor
+        if isinstance(self.grad_op, list):
+            grad_op_l = self.grad_op
+            if len(grad_op_l) > len(self.local_inputs):
+                raise ValueError(
+                    'Can override %d gradients at most, got %d' % (
+                        len(self.local_inputs), len(grad_op_l)))
+            if len(grad_op_l) < len(self.local_inputs):
+                grad_op_l += [None] * (
+                    len(self.local_inputs) - len(grad_op_l))
+            wrt = [self.local_inputs[i] for i, go in
+                   enumerate(grad_op_l) if not go]
+            # compute non-overriding downsteam gradients from upstreams grads
+            # it's normal some input may be disconnected, thus the 'ignore'
+            ups_grads_d = dict(izip(self.local_outputs, output_grads))
+            nat_dns_grads = iter(theano.gradient.grad(
+                cost=None,
+                known_grads=ups_grads_d,
+                wrt=wrt,
+                disconnected_inputs='ignore'))
+            # combine overriding gradients
+            dns_grads_l = [
+                go(self.local_inputs, output_grads) if go else next(nat_dns_grads) for go in grad_op_l]
+            grad_ofg = type(self)(
+                inputs=self.local_inputs + output_grads,
+                outputs=dns_grads_l,
+                inline=self.is_inline, on_unused_input='ignore')
+
+            def grad_op(inps, grds):
+                return grad_ofg(*(list(inps) + list(grds)))
+            self.grad_op = grad_op
```

Comment:
> For my current implementation, there's not way to *ensure* `self.grad_op` is an `OpFromGraph` instance.

Yes, I'm aware. My suggestion was to distinguish that in the case the gradient is already cached:
```python
if not self.grad_op_is_cached:
    ...
    self.grad_op_is_cached = True
if isinstance(self.grad_op, type(self)):
    return self.grad_op(*(inputs, grads))
else:
    return self.grad_op(inputs, grads)
```
To avoid having this distinction both at the top and at the bottom of the `grad()` method, I'm suggesting to change the implementation to first ensure the gradient is cached, then returning the cached gradient (instead of the current implementation of returning the cached gradient if possible, or creating the cache and returning the cached gradient otherwise.)